subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -0,0 +1,181 @@
+        RecompilationSpec spec = new RecompilationSpec();
+
+        processClasspathChanges(current, previous, spec);
+        processOtherChanges(current, previous, spec);
+
+        spec.getClassesToProcess().addAll(previous.getTypesToReprocess());
+        return spec;
+    }
+
+    @Override
+    public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recompilationSpec) {
+        if (!recompilationSpec.isBuildNeeded()) {
+            spec.setSourceFiles(Collections.emptySet());
+            spec.setClasses(Collections.emptySet());
+            return;
+        }
+
+        Factory<PatternSet> patternSetFactory = fileOperations.getFileResolver().getPatternSetFactory();
+        PatternSet classesToDelete = patternSetFactory.create();
+        PatternSet filesToRecompile = patternSetFactory.create();
+
+        prepareFilePatterns(spec, recompilationSpec.getFilesToCompile(), classesToDelete, filesToRecompile);
+
+        spec.setSourceFiles(sourceTree.matching(filesToRecompile));
+        includePreviousCompilationOutputOnClasspath(spec);
+        addClassesToProcess(spec, recompilationSpec);
+
+        deleteStaleFilesIn(classesToDelete, spec.getDestinationDir());
+    }
+
+    private void addClassesToProcess(JavaCompileSpec spec, RecompilationSpec recompilationSpec) {
+        Set<String> classesToProcess = Sets.newHashSet(recompilationSpec.getClassesToProcess());
+        classesToProcess.removeAll(recompilationSpec.getClassesToCompile());
+        spec.setClasses(classesToProcess);
+    }
+
+    private void prepareFilePatterns(JavaCompileSpec spec, Set<File> filesToCompile, PatternSet classesToDelete, PatternSet filesToRecompilePatterns) {
+        for (File file : filesToCompile) {
+            filesToRecompilePatterns.include(relativize(file));
+
+            Collection<String> classes = sourceFileClassNameConverter.getClassNames(file);
+            for (String staleClass : classes) {
+                String path = staleClass.replaceAll("\\.", "/");
+                classesToDelete.include(path.concat(".class"));
+            }
+        }
+    }
+
+    private void includePreviousCompilationOutputOnClasspath(JavaCompileSpec spec) {
+        List<File> classpath = Lists.newArrayList(spec.getCompileClasspath());
+        File destinationDir = spec.getDestinationDir();
+        classpath.add(destinationDir);
+        spec.setCompileClasspath(classpath);
+    }
+
+    private String relativize(File sourceFile) {
+        List<File> dirs = sourceDirs.getSourceRoots();
+        for (File sourceDir : dirs) {
+            if (sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath())) {
+                return RelativePathUtil.relativePath(sourceDir, sourceFile);
+            }
+        }
+        throw new IllegalStateException("Not found " + sourceFile + " in source dirs: " + dirs);
+    }
+
+//    @SuppressWarnings("fallthrough")
```

Comment:
I'm not able to make this code work - some NPEs are thrown and I didn't investigate yet. Currently I'm reusing the alignment code from `JavaRecompilationSpecProvider`.

Reply:
Let's do this in a follow up PR.